### PR TITLE
added scientific notation

### PR DESCRIPTION
**Pull request by minirop**
Hi there, here is my attempt to allow scientific notation of numbers.

### CODE REVIEW

File: src/vm/wren_compiler.c [hunk 1]
```diff
@@ -586,6 +586,21 @@ static void readNumber(Parser* parser)
...
   }
+  
+  // See if the number is in scientific notation
+  if (peekChar(parser) == 'e' || peekChar(parser) == 'E')
+  {
+    nextChar(parser);
+    
+    // if the exponant is negative
+    if(peekChar(parser) == '-')
+      nextChar(parser);
```
Comment (munificent) [inline src/vm/wren_compiler.c]: I hate to make style nits, but can you change this to:

```
if (peekChar(parser) == '-') nextChar(parser);
```

File: src/vm/wren_compiler.c [hunk 2]
```diff
@@ -586,6 +586,21 @@ static void readNumber(Parser* parser)
...
+    
+    if(!isDigit(peekChar(parser)))
+      lexError(parser, "Unterminated scientific notation.");
```
Comment (munificent) [inline src/vm/wren_compiler.c]: And this to:

```c
if (!isDigit(peekChar(parser)))
{
  lexError(parser, "Unterminated scientific notation.");
}
```

My style rules here are:

1. Always have a space after `if` (and other control flow keywords).
2. Only omit the `{}` on an `if` if the entire thing fits on one line.

Thanks!

File: test/language/number/scientific_literals.wren [hunk 3]
```diff
@@ -0,0 +1,9 @@
+var x = 2.55e2
+
+IO.print(x) // expect: 255
+IO.print(x + 1) // expect: 256
+IO.print(x == 255) // expect: true
+IO.print(2.55e-2 is Num) // expect: true
+IO.print(x is Num) // expect: true
+IO.print(-2.55e2) // expect: -255
+IO.print(-25500e-2) // expect: -255
```
Comment (munificent) [inline test/language/number/scientific_literals.wren]: These tests are great!

Can you add a couple of error case tests too? You'll probably need to make each one a separate tiny file. I'm thinking things like:

```
1e
1.e
1.2e
1.2e3e
1.2e3.4
```

Also, two tests like "literal_too_large.wren" to test how it handles an overly long exponent and an overly long mantissa.

I know that's a lot of corner cases, but I think it's really really important to have very detailed test coverage. It's one of the things we can do to give people confidence in embedding Wren in their programs.

**Comment by munificent** [comment]
This is awesome! Thanks for doing this! I have a couple of comments, but otherwise this looks swell. :)

File: test/language/number/scientific_literals.wren (hunk 3 above)
Comment (minirop) [inline test/language/number/scientific_literals.wren]: For the very long exponent and mantissa, I don't really know how to write the tests, since it is the resulting number that counts. If `9999....9999` results in "too large literal", adding `e-100` to it will no longer trigger the error.

**Comment by minirop** [comment]
I've updated the PR with your comments. I've just left out the overly mantissa/exponent (see my other comment).
You (or another linux user) have to "fix" the tests since on windows it's not as flawless (and I'm not really sure what "expect" message I have to put).

File: test/language/number/scientific_missing_fractional_part.wren [hunk 4]
```diff
@@ -0,0 +1 @@
+var x = 1.e // expect error
```
Comment (munificent) [inline test/language/number/scientific_missing_fractional_part.wren]: It looks like this test isn't failing at compile time. It's parsing this as `(1).e`, which makes sense. That allows things like `123.abs` to work so I think the behavior is right here.

You should just be able to change this to expect a runtime error to occur.

Comment (minirop) [inline test/language/number/scientific_missing_fractional_part.wren]: I've just updated the file (I didn't understood that "expect error" was only for compile time errors), but the test still fails. And I have no clue why. The test does pass on my machine.

**Comment by minirop** [comment]
Finally! I found what was failing. It was a '\r' in the test file... :sweat: 
To prevent this, you could remove them on line 79 of *test.py*: `self.runtime_error_message = match.group(1).replace('\r', '')`

File: test/language/number/scientific_literals.wren (hunk 3 above)
Comment (munificent) [inline test/language/number/scientific_literals.wren]: Yeah, that's fine. I don't think we need to worry about this then.

**Comment by munificent** [comment]
This looks great. I'd be happy to merge this in. Before I do, can you sync up (either merge or rebase) against the latest master? I tried to do it locally but wasn't able to get Git into a happy place. The easiest way I could figure out to get your changes on master was to cherry-pick, but I don't want to lose your commit authorship.

**Comment by minirop** [comment]
here we go!

**Comment by munificent** [comment]
Awesome, thank you! You rock!
